SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -29,9 +30,55 @@ variable (q t : Nat) [Fact (q > 1)] (n : Nat)
 
 -- Question: Can we make something like d := 2^n work as a macro?
 --
+theorem WithBot.npow_coe_eq_npow (n : Nat) (x : ℕ) : (WithBot.some x : WithBot ℕ) ^ n = WithBot.some (x ^ n) := by
+  induction n with
+    | zero => simp
+    | succ n ih =>  
+        rw [pow_succ'', ih, ← WithBot.coe_mul]
+        rw [← WithBot.some_eq_coe, WithBot.some]
+        apply Option.some_inj.2
+        rw [Nat.pow_succ]
+        ring
+  done
 
 noncomputable def f : (ZMod q)[X] := X^(2^n) + 1
 
+theorem f_deg_eq : (f q n).degree = 2^n := by
+  simp [f]
+  rw [Polynomial.degree_add_eq_left_of_degree_lt]
+  <;> rw [Polynomial.degree_X_pow]
+  simp
+  simp [Polynomial.degree_one]
+  have h : 0 < 2^n := by
+    apply Nat.one_le_two_pow
+  have h' := WithBot.coe_lt_coe.2 h
+  simp [Preorder.toLT, WithBot.preorder]
+  have h0 : @OfNat.ofNat (WithBot ℕ) 0 Zero.toOfNat0  = @WithBot.some ℕ 0 := by
+    simp [OfNat.ofNat]
+  have h2 : @OfNat.ofNat (WithBot ℕ) 2 instOfNat = @WithBot.some ℕ 2 := by
+    simp [OfNat.ofNat]
+  have h2n : @HPow.hPow (WithBot ℕ) ℕ (WithBot ℕ) instHPow 2 n = @WithBot.some ℕ (@HPow.hPow ℕ ℕ ℕ instHPow 2 n) := by
+    rw [h2, WithBot.npow_coe_eq_npow]
+  rw [h0, h2n]
+  exact h'
+  done
```

Comment:
I am curious, do we need the `done`?

Reply:
no we don't need it, but it's nice! I saw a Zulip thread recommending it (and I wasn't sure if lean was happy with the proof)